packages/contracts/contracts/plugin/PluginSetupProcessor.sol

```diff
@@ -32,27 +33,67 @@ contract PluginSetupProcessor is DaoAuthorizable {
 
-    struct PluginUpdateParams {
-        address plugin; // The address of the `Plugin` contract being currently active. This can be a proxy or a concrete instance.
-        PluginRepo pluginSetupRepo; // The repository storing the `PluginSetup` contracts of all versions of a plugin.
-        address currentPluginSetup; // The `PluginSetup` contract of the version being currently active and to be updated from.
-        address newPluginSetup; // The `PluginSetup` contract of the version to be updated to.
+    /// @notice Used when there's UI update only which gets stored for the setupId generation.
+    /// @dev keccak256(abi.encode[])
+    bytes32 private constant EMPTY_ARRAY_ENCODED_HASH =
+        bytes32(0x569e75fc77c1a856f6daaf9e69d8a9566ca34aa47f9133711ce065a571af0cfd);
+
+    struct PluginInformation {
+        uint256 blockNumber;
+        bytes32 currentSetupId;
+        mapping(bytes32 => uint256) setupIds;
     }
 
-    /// @notice A mapping between a plugin's installations `appliedId` (see [`getAppliedId`](#private-function-`getAppliedId`)) and a boolean expressing if it has been applied. This is used to guarantees that a plugin can only be installed once.
-    mapping(bytes32 => bool) private isInstallationApplied;
+    mapping(bytes32 => PluginInformation) private states;
 
-    /// @notice A mapping between a plugin's [`setupId`](#private-function-`getSetupId`) and the hashed, multi-targeted permission operation list obtained from the [`prepareInstallation`](#external-function-`prepareInstallation`) function.
-    mapping(bytes32 => bytes32) private installPermissionHashes;
+    /// @notice The struct containing the parameters for the `prepareInstallation` function.
+    struct PrepareInstall {
+        PluginSetupRef pluginSetupRef;
+        bytes data;
+    }
 
-    /// @notice A mapping between a plugin's [`setupId`](#private-function-`getSetupId`) and the hashed, multi-targeted permission operation list obtained from the [`prepareUpdate`](#external-function-`prepareUpdate`) function.
-    mapping(bytes32 => bytes32) private updatePermissionHashes;
+    /// @notice The struct containing the parameters for the `applyInstallation` function.
+    struct ApplyInstall {
+        PluginSetupRef pluginSetupRef;
+        address plugin;
+        PermissionLib.ItemMultiTarget[] permissions;
+        bytes32 helpersHash;
+        IDAO.Action[] actions;
+    }
+
+    /// @notice The struct containing the parameters for the `prepareUpdate` function.
+    struct PrepareUpdate {
+        PluginRepo.Tag currentVersionTag;
+        PluginRepo.Tag newVersionTag;
+        PluginRepo pluginSetupRepo;
+        IPluginSetup.SetupPayload setupPayload;
+    }
+
+    /// @notice The struct containing the parameters for the `applyUpdate` function.
+    struct ApplyUpdate {
+        address plugin;
+        PluginSetupRef pluginSetupRef;
+        bytes initData;
+        PermissionLib.ItemMultiTarget[] permissions;
+        bytes32 helpersHash;
```

Comment:
`helpersHash` is the hash of all the helper addresses used by the plugin. The permissions are checked with the passed  `permissions` in combination with the other parameters. This works because all the parameters + the operation hashed together give the `SetupId` and only if the `SetupId` is present the code continues. Please read through the `prepareUpdate` and `applyUpdate` functions to see this struct in action.

To understand the process more easily checkout `prepareInstallation` and `applyInstallation`